File: src/lib/miller.ts
// Client-side Miller-plane geometry for the Viewer node.
// Mirrors atomipy.miller.miller_planes: returns the polygon(s) where the
// (h,k,l) lattice plane(s) intersect the cell, as ordered Cartesian vertices.

export type Cell = { a: number; b: number; c: number; alpha: number; beta: number; gamma: number };
export type Vec3 = [number, number, number];

const VFRAC: Vec3[] = [
  [0, 0, 0], [1, 0, 0], [0, 1, 0], [1, 1, 0],
  [0, 0, 1], [1, 0, 1], [0, 1, 1], [1, 1, 1],
];
const EDGES: [number, number][] = [
  [0, 1], [2, 3], [4, 5], [6, 7],
  [0, 2], [1, 3], [4, 6], [5, 7],
  [0, 4], [1, 5], [2, 6], [3, 7],
];

/** Parse the cell (a,b,c,α,β,γ) from a PDB CRYST1 record. Null if absent/invalid. */
export function parseCryst1(pdb: string): Cell | null {
  const m = pdb.match(/^CRYST1.*$/m);
  if (!m) return null;
  const line = m[0];
  // PDB fixed columns first; fall back to whitespace split.
  let a = parseFloat(line.slice(6, 15));
  let b = parseFloat(line.slice(15, 24));
  let c = parseFloat(line.slice(24, 33));
  let alpha = parseFloat(line.slice(33, 40));
  let beta = parseFloat(line.slice(40, 47));
  let gamma = parseFloat(line.slice(47, 54));
  const bad = (vals: number[]) => vals.some((v) => !Number.isFinite(v) || v <= 0);
  if (bad([a, b, c, alpha, beta, gamma])) {
    const p = line.trim().split(/\s+/);
    a = parseFloat(p[1]); b = parseFloat(p[2]); c = parseFloat(p[3]);
    alpha = parseFloat(p[4]); beta = parseFloat(p[5]); gamma = parseFloat(p[6]);
    if (bad([a, b, c, alpha, beta, gamma])) return null;
  }
  return { a, b, c, alpha, beta, gamma };
}

/** 3x3 fractional→Cartesian matrix (columns a, b, c). */
function fromFracMatrix(cell: Cell): number[][] {
  const d2r = Math.PI / 180;
  const al = cell.alpha * d2r, be = cell.beta * d2r, ga = cell.gamma * d2r;
  const ca = Math.cos(al), cb = Math.cos(be), cg = Math.cos(ga), sg = Math.sin(ga);
  const v = Math.sqrt(Math.max(0, 1 - ca * ca - cb * cb - cg * cg + 2 * ca * cb * cg));
  return [
    [cell.a, cell.b * cg, cell.c * cb],
    [0, cell.b * sg, cell.c * (ca - cb * cg) / sg],
    [0, 0, cell.c * v / sg],
  ];
}

function matVec(M: number[][], x: Vec3): Vec3 {
  return [
    M[0][0] * x[0] + M[0][1] * x[1] + M[0][2] * x[2],
    M[1][0] * x[0] + M[1][1] * x[1] + M[1][2] * x[2],
    M[2][0] * x[0] + M[2][1] * x[1] + M[2][2] * x[2],
  ];
}
const sub = (a: Vec3, b: Vec3): Vec3 => [a[0] - b[0], a[1] - b[1], a[2] - b[2]];
const cross = (a: Vec3, b: Vec3): Vec3 => [
  a[1] * b[2] - a[2] * b[1], a[2] * b[0] - a[0] * b[2], a[0] * b[1] - a[1] * b[0],
];
const dot = (a: Vec3, b: Vec3) => a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
const norm = (a: Vec3) => Math.sqrt(dot(a, a));

/** Interplanar spacing d_hkl (Å). */
export function dSpacing(h: number, k: number, l: number, cell: Cell): number {
  const M = fromFracMatrix(cell);
  // real-space metric G = MᵀM
  const G = [0, 1, 2].map((i) => [0, 1, 2].map((j) => M[0][i] * M[0][j] + M[1][i] * M[1][j] + M[2][i] * M[2][j]));
  // inv(G) (symmetric 3x3)
  const det =
    G[0][0] * (G[1][1] * G[2][2] - G[1][2] * G[2][1]) -
    G[0][1] * (G[1][0] * G[2][2] - G[1][2] * G[2][0]) +
    G[0][2] * (G[1][0] * G[2][1] - G[1][1] * G[2][0]);
  if (Math.abs(det) < 1e-30) return Infinity;
  const inv = [
    [(G[1][1] * G[2][2] - G[1][2] * G[2][1]) / det, (G[0][2] * G[2][1] - G[0][1] * G[2][2]) / det, (G[0][1] * G[1][2] - G[0][2] * G[1][1]) / det],
    [(G[1][2] * G[2][0] - G[1][0] * G[2][2]) / det, (G[0][0] * G[2][2] - G[0][2] * G[2][0]) / det, (G[0][2] * G[1][0] - G[0][0] * G[1][2]) / det],
    [(G[1][0] * G[2][1] - G[1][1] * G[2][0]) / det, (G[0][1] * G[2][0] - G[0][0] * G[2][1]) / det, (G[0][0] * G[1][1] - G[0][1] * G[1][0]) / det],
  ];
  const hkl = [h, k, l];
  let invd2 = 0;
  for (let i = 0; i < 3; i++) for (let j = 0; j < 3; j++) invd2 += hkl[i] * inv[i][j] * hkl[j];
  return invd2 > 0 ? 1 / Math.sqrt(invd2) : Infinity;
}

function uniqueRows(pts: Vec3[], tol = 1e-9): Vec3[] {
  const out: Vec3[] = [];
  for (const p of pts) {
    if (!out.some((o) => Math.abs(o[0] - p[0]) < tol && Math.abs(o[1] - p[1]) < tol && Math.abs(o[2] - p[2]) < tol)) {
      out.push(p);
    }
  }
  return out;
}

function intersectPolygon(M: number[][], h: number, k: number, l: number, s: number): Vec3[] | null {
  const pts: Vec3[] = [];
  for (const [e0, e1] of EDGES) {
    const v0 = VFRAC[e0], v1 = VFRAC[e1];
    const dv: Vec3 = [v1[0] - v0[0], v1[1] - v0[1], v1[2] - v0[2]];
    const denom = h * dv[0] + k * dv[1] + l * dv[2];
    if (Math.abs(denom) < 1e-14) continue;
    const t = (s - (h * v0[0] + k * v0[1] + l * v0[2])) / denom;
    if (t < -1e-12 || t > 1 + 1e-12) continue;
    const p: Vec3 = [v0[0] + t * dv[0], v0[1] + t * dv[1], v0[2] + t * dv[2]];
    if (p.every((q) => q >= -1e-12 && q <= 1 + 1e-12)) {
      pts.push([Math.min(1, Math.max(0, p[0])), Math.min(1, Math.max(0, p[1])), Math.min(1, Math.max(0, p[2]))]);
    }
  }
  const P = uniqueRows(pts);
  if (P.length < 3) return null;
  // order in-plane by angle around centroid
  const c0: Vec3 = [0, 0, 0];
  P.forEach((p) => { c0[0] += p[0]; c0[1] += p[1]; c0[2] += p[2]; });
  c0[0] /= P.length; c0[1] /= P.length; c0[2] /= P.length;
  const nvec: Vec3 = [h, k, l];
  let ref: Vec3 = [1, 0, 0];
  if (norm(cross(nvec, ref)) < 1e-12) ref = [0, 1, 0];
  if (norm(cross(nvec, ref)) < 1e-12) ref = [0, 0, 1];
  let u = cross(nvec, ref);
  const nu = norm(u);
  u = [u[0] / nu, u[1] / nu, u[2] / nu];
  let w = cross(nvec, u);
  const nw = norm(w);
  w = [w[0] / nw, w[1] / nw, w[2] / nw];
  P.sort((p, q) => {
    const ap = sub(p, c0), aq = sub(q, c0);
    return Math.atan2(dot(ap, w), dot(ap, u)) - Math.atan2(dot(aq, w), dot(aq, u));
  });
  return P.map((p) => matVec(M, p)); // fractional → Cartesian
}

export type MillerOpts = {
  singlePlane?: boolean;   // false (default) = full family of planes
  planeLevel?: number | "auto";
  offset?: number;         // shift along normal, in Å
};

/** Polygons (Cartesian vertices) where the (h,k,l) plane(s) cut the cell. */
export function millerPlanes(h: number, k: number, l: number, cell: Cell, opts: MillerOpts = {}): Vec3[][] {
  h = Math.round(h); k = Math.round(k); l = Math.round(l);
  if (h === 0 && k === 0 && l === 0) return [];
  const M = fromFracMatrix(cell);
  const offset = opts.offset || 0;
  const ds = offset ? offset / dSpacing(h, k, l, cell) : 0;
  const svals = VFRAC.map((v) => h * v[0] + k * v[1] + l * v[2]);
  const nmin = Math.ceil(Math.min(...svals) - ds);
  const nmax = Math.floor(Math.max(...svals) - ds);
  const polys: Vec3[][] = [];
  if (opts.singlePlane) {
    let levels: number[];
    if (opts.planeLevel === undefined || opts.planeLevel === "auto") {
      levels = [];
      for (let n = nmin; n <= nmax; n++) levels.push(n);
      levels.sort((p, q) => Math.abs(p - 1) - Math.abs(q - 1));
    } else {
      // Explicit fractional level (may be non-integer, e.g. the structure
      // midpoint) — use it as-is, do NOT round.
      levels = [opts.planeLevel as number];
    }
    for (const n of levels) {
      const poly = intersectPolygon(M, h, k, l, n + ds);
      if (poly) { polys.push(poly); break; }
    }
  } else {
    for (let n = nmin; n <= nmax; n++) {
      const poly = intersectPolygon(M, h, k, l, n + ds);
      if (poly) polys.push(poly);
    }
  }
  return polys;
}

/** Fan-triangulate a polygon into a 3Dmol addCustom spec (double-sided). */
export function polygonTo3Dmol(poly: Vec3[]): { vertexArr: { x: number; y: number; z: number }[]; normalArr: { x: number; y: number; z: number }[]; faceArr: number[] } {
  const vertexArr = poly.map((p) => ({ x: p[0], y: p[1], z: p[2] }));
  let n = cross(sub(poly[1], poly[0]), sub(poly[2], poly[0]));
  const nn = norm(n) || 1;
  n = [n[0] / nn, n[1] / nn, n[2] / nn];
  const normalArr = poly.map(() => ({ x: n[0], y: n[1], z: n[2] }));
  const faceArr: number[] = [];
  for (let i = 1; i < poly.length - 1; i++) {
    faceArr.push(0, i, i + 1);     // front
    faceArr.push(0, i + 1, i);     // back (so it's visible from both sides)
  }
  return { vertexArr, normalArr, faceArr };
}

/** "Auto" plane level = midpoint of the structure along the (h,k,l) normal,
 *  in fractional plane-level units — the SAME convention the Edit cut uses, so
 *  the overlay plane lands exactly where the cut would. Parses atom coordinates
 *  from a PDB string. Returns null if no atoms/cell. */
export function atomMidLevel(pdb: string, cell: Cell, h: number, k: number, l: number): number | null {
  const M = fromFracMatrix(cell);
  // invert M (3x3) to map Cartesian -> fractional
  const det =
    M[0][0] * (M[1][1] * M[2][2] - M[1][2] * M[2][1]) -
    M[0][1] * (M[1][0] * M[2][2] - M[1][2] * M[2][0]) +
    M[0][2] * (M[1][0] * M[2][1] - M[1][1] * M[2][0]);
  if (Math.abs(det) < 1e-30) return null;
  const inv = [
    [(M[1][1] * M[2][2] - M[1][2] * M[2][1]) / det, (M[0][2] * M[2][1] - M[0][1] * M[2][2]) / det, (M[0][1] * M[1][2] - M[0][2] * M[1][1]) / det],
    [(M[1][2] * M[2][0] - M[1][0] * M[2][2]) / det, (M[0][0] * M[2][2] - M[0][2] * M[2][0]) / det, (M[0][2] * M[1][0] - M[0][0] * M[1][2]) / det],
    [(M[1][0] * M[2][1] - M[1][1] * M[2][0]) / det, (M[0][1] * M[2][0] - M[0][0] * M[2][1]) / det, (M[0][0] * M[1][1] - M[0][1] * M[1][0]) / det],
  ];
  let fmin = Infinity, fmax = -Infinity, n = 0;
  for (const line of pdb.split("\n")) {
    if (!(line.startsWith("ATOM") || line.startsWith("HETATM"))) continue;
    const x = parseFloat(line.slice(30, 38)), y = parseFloat(line.slice(38, 46)), z = parseFloat(line.slice(46, 54));
    if (!Number.isFinite(x) || !Number.isFinite(y) || !Number.isFinite(z)) continue;
    const xf = inv[0][0] * x + inv[0][1] * y + inv[0][2] * z;
    const yf = inv[1][0] * x + inv[1][1] * y + inv[1][2] * z;
    const zf = inv[2][0] * x + inv[2][1] * y + inv[2][2] * z;
    const f = h * xf + k * yf + l * zf;
    if (f < fmin) fmin = f;
    if (f > fmax) fmax = f;
    n++;
  }
  // No atoms, or a degenerate set (e.g. just the dummy atom for an empty box,
  // or a single atomic layer) → return null so the caller falls back to the
  // cell-based "auto" level and still draws a plane across the box.
  if (n === 0 || !(fmax > fmin + 1e-9)) return null;
  return 0.5 * (fmin + fmax);
}

function hexToRgb(hex: string): [number, number, number] {
  const m = /^#?([0-9a-f]{6})$/i.exec(hex.trim());
  if (!m) return [245, 158, 11]; // amber fallback
  const v = parseInt(m[1], 16);
  return [(v >> 16) & 255, (v >> 8) & 255, v & 255];
}

/** Jmol `draw POLYGON` commands for one plane's polygons (translucent).
 *  `idPrefix` namespaces the draw IDs so multiple planes don't collide.
 *  (Caller is responsible for an initial `draw miller* delete` to clear.) */
export function jmolMillerCommands(polys: Vec3[][], hexColor: string, opacity: number, idPrefix = "miller"): string[] {
  const [r, g, b] = hexToRgb(hexColor);
  const translucent = Math.min(1, Math.max(0, 1 - opacity)); // Jmol: 0 opaque .. 1 transparent
  const cmds: string[] = [];
  polys.forEach((poly, idx) => {
    const verts = poly.map((p) => `{${p[0].toFixed(4)} ${p[1].toFixed(4)} ${p[2].toFixed(4)}}`).join(" ");
    const tris: string[] = [];
    for (let i = 1; i < poly.length - 1; i++) tris.push(`[0 ${i} ${i + 1}]`);
    const id = `${idPrefix}${idx}`;
    cmds.push(`draw ${id} POLYGON ${poly.length} ${verts} ${tris.length} ${tris.join(" ")}`);
    cmds.push(`color $${id} translucent ${translucent.toFixed(2)} [${r} ${g} ${b}]`);
  });
  return cmds;
}

File: atomipy/miller.py
"""Miller-plane geometry.

Compute the polygon(s) where a set of (h, k, l) lattice planes intersect the
simulation cell — useful for visualising crystallographic planes on top of a
structure. Pure geometry (no plotting); ported from the MATLAB atom toolbox
``show_miller.m``.

Main entry point: :func:`miller_planes`. Helper: :func:`d_spacing`.
"""

import numpy as np

from .cell_utils import Box_dim2Cell

# 8 corners of the fractional unit cube and its 12 edges (0-based vertex indices).
_VFRAC = np.array(
    [
        [0, 0, 0], [1, 0, 0], [0, 1, 0], [1, 1, 0],
        [0, 0, 1], [1, 0, 1], [0, 1, 1], [1, 1, 1],
    ],
    dtype=float,
)
_EDGES = np.array(
    [
        [0, 1], [2, 3], [4, 5], [6, 7],
        [0, 2], [1, 3], [4, 6], [5, 7],
        [0, 4], [1, 5], [2, 6], [3, 7],
    ]
)


def _cell_from_box(Box):
    """Return (a, b, c, alpha, beta, gamma) [deg] from a 3/6/9-element Box."""
    box = list(np.asarray(Box, dtype=float).ravel())
    n = len(box)
    if n == 3:
        a, b, c = box
        return a, b, c, 90.0, 90.0, 90.0
    if n == 6:
        return tuple(box)
    if n == 9:
        cell = Box_dim2Cell(box)
        return tuple(float(x) for x in cell[:6])
    raise ValueError("Box must have 3, 6, or 9 elements.")


def _from_frac_matrix(Box):
    """3x3 matrix mapping fractional -> Cartesian coordinates (columns a, b, c)."""
    a, b, c, al, be, ga = _cell_from_box(Box)
    al, be, ga = np.radians([al, be, ga])
    v = np.sqrt(
        max(
            0.0,
            1 - np.cos(al) ** 2 - np.cos(be) ** 2 - np.cos(ga) ** 2
            + 2 * np.cos(al) * np.cos(be) * np.cos(ga),
        )
    )
    return np.array(
        [
            [a, b * np.cos(ga), c * np.cos(be)],
            [0, b * np.sin(ga), c * (np.cos(al) - np.cos(be) * np.cos(ga)) / np.sin(ga)],
            [0, 0, c * v / np.sin(ga)],
        ]
    )


def d_spacing(h, k, l, Box):
    """Interplanar spacing d_hkl (Angstrom) for the (h, k, l) family.

    Computed from the real-space metric tensor G = MᵀM (M = fractional→Cartesian
    matrix) as 1/d² = [h k l] · G⁻¹ · [h k l]ᵀ.
    """
    M = _from_frac_matrix(Box)
    G = M.T @ M
    hkl = np.array([h, k, l], dtype=float)
    inv_d2 = float(hkl @ np.linalg.inv(G) @ hkl)
    if inv_d2 <= 0:
        return float("inf")
    return 1.0 / np.sqrt(inv_d2)


def _unique_rows(arr, tol=1e-10):
    out = []
    for row in arr:
        if not any(np.allclose(row, o, atol=tol) for o in out):
            out.append(row)
    return np.array(out)


def _polygon_area3d(P):
    c0 = P.mean(axis=0)
    acc = np.zeros(3)
    m = len(P)
    for i in range(m):
        acc = acc + np.cross(P[i] - c0, P[(i + 1) % m] - c0)
    return 0.5 * float(np.linalg.norm(acc))


def _intersect_polygon(FromFrac, hkl, s, exclude_boundary):
    """Ordered Cartesian polygon where plane h·x+k·y+l·z = s cuts the unit cube."""
    h, k, l = hkl
    pts = []
    for e0, e1 in _EDGES:
        v0 = _VFRAC[e0]
        v1 = _VFRAC[e1]
        dv = v1 - v0
        denom = h * dv[0] + k * dv[1] + l * dv[2]
        if abs(denom) < 1e-14:
            continue
        t = (s - (h * v0[0] + k * v0[1] + l * v0[2])) / denom
        if -1e-12 <= t <= 1 + 1e-12:
            p = v0 + t * dv
            if np.all(p >= -1e-12) and np.all(p <= 1 + 1e-12):
                pts.append(np.clip(p, 0.0, 1.0))
    if len(pts) < 3:
        return None
    P = _unique_rows(np.array(pts))
    if P.shape[0] < 3:
        return None

    # Order the vertices in-plane by angle around their centroid.
    c0 = P.mean(axis=0)
    nvec = np.array(hkl, dtype=float)
    ref = np.array([1.0, 0.0, 0.0])
    if np.linalg.norm(np.cross(nvec, ref)) < 1e-12:
        ref = np.array([0.0, 1.0, 0.0])
    if np.linalg.norm(np.cross(nvec, ref)) < 1e-12:
        ref = np.array([0.0, 0.0, 1.0])
    u = np.cross(nvec, ref)
    nu = np.linalg.norm(u)
    if nu < 1e-14:
        _, _, Vt = np.linalg.svd(P - c0)
        u, w = Vt[0], Vt[1]
    else:
        u = u / nu
        w = np.cross(nvec, u)
        w = w / np.linalg.norm(w)
    ang = np.arctan2((P - c0) @ w, (P - c0) @ u)
    P = P[np.argsort(ang)]

    if exclude_boundary and _polygon_area3d(P) <= 1e-12:
        return None
    return (FromFrac @ P.T).T  # fractional -> Cartesian


def cut_planes(atoms, Box, planes, whole_molecules=False, reindex=True):
    """Keep only atoms that satisfy ALL given Miller half-spaces (intersection).

    Each plane carves a half-space; intersecting several carves a convex region
    — e.g. 6 side planes (60° apart) cut a **hexagonal column** out of a clay
    layer, optionally capped by 2 basal planes. In fractional coordinates each
    plane is a linear threshold f = h·xf + k·yf + l·zf = s, so the whole cut is
    a logical AND of comparisons (works for any (hkl) and triclinic cell).

    Parameters
    ----------
    atoms : list of dict
        Atom records with 'x', 'y', 'z' (Cartesian, Angstrom).
    Box : sequence
        3/6/9-element cell (see :func:`miller_planes`).
    planes : list of dict
        Each plane: ``{'h','k','l', 'side': 'below'|'above', 'level': 'auto'|float,
        'offset': float}``. 'below' keeps f <= s, 'above' keeps f >= s; 'auto'
        level = midpoint of the structure along that normal; offset shifts the
        plane along its normal in Angstrom (Δs = offset / d_hkl).
    whole_molecules : bool
        Keep/drop whole molecules by centroid (uses 'molid') so molecules aren't
        sliced.
    reindex : bool
        Renumber surviving atoms' 'index' 1..N (default True).

    Returns
    -------
    list of dict
        Surviving atoms (new list; originals untouched).
    """
    if not atoms:
        return []
    if not planes:
        return [dict(a) for a in atoms]

    M = _from_frac_matrix(Box)
    Minv = np.linalg.inv(M)
    cart = np.array([[float(a.get("x", 0.0)), float(a.get("y", 0.0)), float(a.get("z", 0.0))] for a in atoms])
    frac = cart @ Minv.T
    tol = 1e-9  # atoms exactly on a plane (and cos(90°)≈1e-16 noise) stay on the kept side
    molids = np.array([a.get("molid", i) for i, a in enumerate(atoms)]) if whole_molecules else None

    mask = np.ones(len(atoms), dtype=bool)
    for pl in planes:
        h, k, l = int(round(pl["h"])), int(round(pl["k"])), int(round(pl["l"]))
        if h == 0 and k == 0 and l == 0:
            continue
        hkl = np.array([h, k, l], dtype=float)
        f = frac @ hkl
        level = pl.get("level", "auto")
        s = 0.5 * (float(f.min()) + float(f.max())) if isinstance(level, str) else float(level)
        off = pl.get("offset", 0.0)
        if off:
            d = d_spacing(h, k, l, Box)
            if np.isfinite(d) and d > 0:
                s += off / d
        side = pl.get("side", "below")
        if whole_molecules:
            pm = np.zeros(len(atoms), dtype=bool)
            for mid in np.unique(molids):
                sel = molids == mid
                mf = float(f[sel].mean())
                pm[sel] = (mf <= s + tol) if side == "below" else (mf >= s - tol)
            mask &= pm
        else:
            mask &= (f <= s + tol) if side == "below" else (f >= s - tol)

    kept = [dict(a) for a, m in zip(atoms, mask) if m]
    if reindex:
        for i, a in enumerate(kept):
            a["index"] = i + 1
    return kept


def cut_miller(atoms, Box, h, k, l, level="auto", offset=0.0, side="below",
               whole_molecules=False, reindex=True):
    """Keep atoms on one side of a single (h, k, l) Miller plane.

    Convenience wrapper around :func:`cut_planes` for the common single-plane
    case. 'below' keeps the inner side (f <= s), 'above' the outer side. See
    :func:`cut_planes` for cutting a convex region with several planes.
    """
    if int(round(h)) == 0 and int(round(k)) == 0 and int(round(l)) == 0:
        raise ValueError("At least one of h, k, l must be non-zero.")
    return cut_planes(
        atoms, Box,
        [{"h": h, "k": k, "l": l, "side": side, "level": level, "offset": offset}],
        whole_molecules=whole_molecules, reindex=reindex,
    )


def miller_planes(
    h,
    k,
    l,
    Box=(1, 1, 1),
    single_plane=False,
    plane_level="auto",
    offset=0.0,
    exclude_boundary=False,
):
    """Polygon(s) where the (h, k, l) Miller plane(s) intersect the cell.

    Parameters
    ----------
    h, k, l : int
        Miller indices (any sign combination; at least one must be non-zero).
    Box : sequence, optional
        Cell as 3 ``[a, b, c]``, 6 ``[a, b, c, α, β, γ]``, or 9
        ``[lx, ly, lz, 0, 0, xy, 0, xz, yz]`` elements. Default unit cube.
    single_plane : bool, optional
        If False (default), return the **full family** of parallel (hkl) planes
        that intersect the cell. If True, return a single plane (at
        ``plane_level``).
    plane_level : int or 'auto', optional
        For ``single_plane=True``: the integer level n of the plane
        ``h·x + k·y + l·z = n`` (fractional coords). ``'auto'`` (default) picks
        the in-cell level nearest to 1.
    offset : float, optional
        Shift the plane(s) along their normal, in Angstrom. Converted to a
        fractional-level shift Δ = offset / d_hkl, so ``offset = d_spacing(...)``
        moves the plane(s) by exactly one interplanar spacing. Default 0.
    exclude_boundary : bool, optional
        Drop degenerate (zero-area, edge/vertex-only) contacts. Default False.

    Returns
    -------
    list of numpy.ndarray
        One (N×3) array of ordered Cartesian vertices per plane polygon; empty
        if no plane intersects the cell.
    """
    h, k, l = int(round(h)), int(round(k)), int(round(l))
    if h == 0 and k == 0 and l == 0:
        raise ValueError("At least one of h, k, l must be non-zero.")

    FromFrac = _from_frac_matrix(Box)
    hkl = (h, k, l)

    # Convert an Angstrom offset along the normal to a fractional-level shift.
    ds = 0.0
    if offset:
        d = d_spacing(h, k, l, Box)
        ds = offset / d if np.isfinite(d) and d > 0 else 0.0

    svals = _VFRAC @ np.array([h, k, l], dtype=float)
    nmin = int(np.ceil(svals.min() - ds))
    nmax = int(np.floor(svals.max() - ds))

    polys = []
    if single_plane:
        if isinstance(plane_level, str):  # 'auto' -> nearest usable level to 1
            levels = sorted(range(nmin, nmax + 1), key=lambda n: abs(n - 1))
        else:
            levels = [int(round(plane_level))]
        for n in levels:
            poly = _intersect_polygon(FromFrac, hkl, n + ds, exclude_boundary)
            if poly is not None:
                polys.append(poly)
                break
    else:
        for n in range(nmin, nmax + 1):
            poly = _intersect_polygon(FromFrac, hkl, n + ds, exclude_boundary)
            if poly is not None:
                polys.append(poly)
    return polys
